app/src/main/java/io/appium/uiautomator2/model/XPathFinder.java

```diff
@@ -51,235 +49,200 @@
-      new HashMap<Element, UiElement<?, ?>>();
+    private static final XPath XPATH_COMPILER = XPathFactory.newInstance().newXPath();
+    private static final String UI_ELEMENT_INDEX = "uiElementIndex";
+    private final String xPathString;
 
-  public static void clearData() {
-    TO_DOM_MAP.clear();
-    FROM_DOM_MAP.clear();
-    document = null;
-  }
-
-  private final String xPathString;
-  private final XPathExpression xPathExpression;
-  private static UiAutomationElement rootElement;
-
-  @Override
-  public String toString() {
-    return xPathString;
-  }
-
-  public XPathFinder(String xPathString) {
-    this.xPathString = Preconditions.checkNotNull(xPathString);
-    try {
-      xPathExpression = XPATH_COMPILER.compile(xPathString);
-    } catch (XPathExpressionException e) {
-      throw new UiAutomator2Exception("xPathString=" + xPathString, e);
+    @Override
+    public String toString() {
+        return xPathString;
     }
-  }
 
-  @Override
-  public NodeInfoList find(UiElement context) {
-    Element domNode = getDomNode((UiElement<?, ?>) context);
-    try {
-      getDocument().appendChild(domNode);
-      NodeList nodes = (NodeList) xPathExpression.evaluate(domNode, XPathConstants.NODESET);
-      NodeInfoList list = new NodeInfoList();
+    private XPathFinder(String xPathString) {
+        this.xPathString = Preconditions.checkNotNull(xPathString);
+    }
 
-      int nodesLength = nodes.getLength();
-        for (int i = 0; i < nodesLength; i++) {
-          if (nodes.item(i).getNodeType() == Node.ELEMENT_NODE && !FROM_DOM_MAP.get(nodes.item(i)).getClassName().equals("hierarchy")) {
-              list.addToList(FROM_DOM_MAP.get(nodes.item(i)).node);
-          }
+    @Override
+    public NodeInfoList find(UiElement context) {
+        final Document document;
+        try {
+            document = DocumentBuilderFactory.newInstance()
+                    .newDocumentBuilder()
+                    .newDocument();
+        } catch (ParserConfigurationException e) {
+            throw new UiAutomator2Exception(e);
         }
-      return list;
-    } catch (XPathExpressionException e) {
-      throw new ElementNotFoundException( e);
-    } finally {
-      try {
-        getDocument().removeChild(domNode);
-      } catch (DOMException e) {
-        Logger.error(e, "Failed to clear document");
-        document = null; // getDocument will create new
-      }
+        final SparseArray<UiElement<?, ?>> uiElementsMapping = new SparseArray<>();
+        final Element domNode = toDOMElement((UiElement<?, ?>) context, document, uiElementsMapping);
+        document.appendChild(domNode);
+        final NodeList nodes;
+        try {
+            nodes = (NodeList) XPATH_COMPILER
+                    .compile(xPathString)
+                    .evaluate(domNode, XPathConstants.NODESET);
+        } catch (XPathExpressionException e) {
+            throw new ElementNotFoundException(e);
```

Comment:
Maybe it would be better to throw  InvalidSelectorException so the handlers can return response with status WDStatus.INVALID_SELECTOR instead of NO_SUCH_ELEMENT.

Reply:
changed. also configured all custom exceptions to be inherited from UiAutomator2Exception